saito-core/src/core/util/rate_limiter.rs

```diff
@@ -0,0 +1,49 @@
+use crate::core::defs::Timestamp;
+
+pub enum RateLimiterRequestType {
+    KeyList,
+    HandshakeChallenge,
+}
+
+#[derive(Debug, Clone)]
+pub struct RateLimiter {
+    limit: usize,                 // Max allowed requests in the window
+    window: Timestamp,            // Window duration in milliseconds
+    request_count: usize,         // Number of requests made in the current window
+    last_request_time: Timestamp, // Timestamp of the last request in milliseconds
+}
+
+impl RateLimiter {
+    pub fn set_limit(&mut self, limit: usize) {
+        self.limit = limit;
+    }
+
+    pub fn set_window(&mut self, window: u64) {
+        self.window = window;
+    }
+
+    pub fn can_make_request(&mut self, current_time: u64) -> bool {
+        if current_time.saturating_sub(self.last_request_time) > self.window {
+            self.request_count = 0;
+            self.last_request_time = current_time;
+        }
```

Comment:
should we return here ?

Reply:
No we should not, because we still need to allow the current request to be processed